chore: sitemaps api handler (#2)
moxystudio/next-sitemaps: README.md

[codecov-url]:https://codecov.io/gh/moxystudio/next-sitemaps-plugin
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/next-sitemaps-plugin/master.svg

> This module provides an API handler and a plugin that enables your next.js application to have working sitemaps so that search engine crawlers can index all your pages.

Reviewer: ```suggestion
This module provides an API handler and a plugin that enables your Next.js application to have working sitemaps so that search engine crawlers can index all your pages.
```

## Installation

```cmd
npm install @moxy/next-sitemaps
```

## Usage

### Simple usage

**1. Add the plugin to `next.config.js`**

```js

> const { withSitemap } = require('@moxy/next-sitemaps');

Reviewer: We must change this to:

```js
const withSitemap = require('@moxy/next-sitemaps/plugin');
```

See below.

withPlugins([
    //other plugins..
    withSitemap()
    //other plugins...

> ])

Reviewer: ```suggestion
]);
```

```

**2. Add the API handler on `pages/api/sitemap.js`**

```js

> import { setupSitemap } from '@moxy/next-sitemaps';

Reviewer: Because the plugin is no longer exported, we may now use the default export like so:

```js
import setupSitemap from '@moxy/next-sitemaps';
```

and perhaps a better name would be:

```js
import createSitemapApiHandler from '@moxy/next-sitemaps';
// or..
import createApiHandler from '@moxy/next-sitemaps';
```

export default setupSitemap({
    baseUrl: 'https://moxy.studio',
    mapDynamicRoutes: {
        '/[id]': () => ['id1', 'id2'],
    },
});
```

> 

Reviewer: I think we are missing instructions to add `robots.txt` with the sitemap URL, see: https://support.google.com/webmasters/answer/183668?hl=en

Reviewer: Please note that the URL must be a "full" URL in `robots.txt`, just like the `baseURL` option.

## API

### withSitemap()

> This plugin will match all files and directories inside the next's `/pages` folder and it will map them into urls. Then, a global named `__NEXT_ROUTES__` will be created with all the urls.

Reviewer: ```suggestion
This plugin will match all files and directories inside the `/pages` folder and it will map them into URLs. Those mappings will be injected in a global variable called  `__NEXT_ROUTES__`, which will be picked by the API handler.
```

> > ℹ️ If you are on dev environment and you have added a new route, you must restart your server to update the global `__NEXT_ROUTES__` and hence, get the correct sitemap.

Reviewer: ```suggestion
> ℹ️ During development. you must restart your server the server whenever you add, remove or modify an existing page, otherwise it won't be reflected in the `/sitemap` response.
```

Author: Your suggestion is a little bit confusing, but I got the idea. 
I think `modify` can confuse developers. You don't need to restart your server when you modify your pages' UI. You just need to restart it if you rename it. 

See my suggestion below and let me know what you think.

> During development, you must restart the server whenever you add, remove or rename an existing page. Otherwise, it won't be reflected in the `/sitemap` response.

Author: In fact, this is not true anymore considering this new approach. I'll remove this side note.

### setupSitemap([options])

> Defines an API handler that will send a mapped sitemap.xml.

Reviewer: ```suggestion
Defines an API handler that will respond with a valid sitemap XML file containing the website pages.
```

#### options

Type: `object`

##### baseUrl

> Type: `string`

Reviewer: ```suggestion
Type: `string`   
```
So that they are in separate lines.

Default: `'/'`

> The website url to join on each sitemap entry. [More info](#specifying-the-base-url).

Reviewer: ```suggestion
The website base URL, which will be used to prefix all page pathnames. Please check [Specifying the base URL](#specifying-the-base-url) for more info.
```

##### mapDynamicRoutes

> Type: `object`

Reviewer: ```suggestion
Type: `object`   
```

Default: `{}`

An object that indicates the possible values for each dynamic route. [More info](#handling-dynamic-routing).

##### handleWarning

> Type: `function`

Reviewer: ```suggestion
Type: `function`   
```

Default: see `*handleWarningDefault*` in [index.js](./index.js)

A function that handle possible warnings. It has the following signature: `(message) => {}`. [More info](#specifying-a-custom-warning-function).

> ##### handleError

Reviewer: I would simplify this and I would not allow customize the response. Bots don't care about the response body, they just ignore anything that is different than 2xx. What I'm saying is to change this to `logError`, just like https://github.com/moxystudio/next-rest-api#logerror. This way we keep these options consistent across packages.

The `handleWarning` would also be renamed to `logWarning`.

If you do this, please adjust the parameters descriptions. Also remove the "More Info" has it feels unnecessary to further document something that is straightforward.

> Type: `function`

Reviewer: ```suggestion
Type: `function`   
```

Default: see `*handleErrorDefault*` in [index.js](./index.js)

A function that handle possible errors. It has the following signature: `(err) => {}`. [More info](#specifying-a-custom-error-function).

### Specifying the base url

```js
setupSitemap({
    baseUrl: 'https://moxy.studio'

> }),

Reviewer: ```suggestion
});
```

Reviewer: Do the same in all examples?

[codecov-url]:https://codecov.io/gh/moxystudio/next-sitemaps-plugin
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/next-sitemaps-plugin/master.svg

This module provides an API handler and a plugin that enables your Next.js application to have working sitemaps so that search engine crawlers can index all your pages.

## Installation

```cmd
npm install @moxy/next-sitemaps
```

## Usage

### Simple usage

**1. Add the plugin to `next.config.js`**

```js
const withSitemap = require('@moxy/next-sitemaps/plugin');

withPlugins([
    //other plugins..
    withSitemap()
    //other plugins...
]);
```

**2. Add the API handler on `pages/api/sitemap.js`**

```js
import createSitemapApiHandler from '@moxy/next-sitemaps';

export default createSitemapApiHandler({
    baseUrl: 'https://moxy.studio',
    mapDynamicRoutes: {
        '/[id]': () => ['id1', 'id2'],
    },
});
```

**3. Add the endpoint URL to your project's robots.txt file**

```txt
Sitemap: https://moxy.studio/api/sitemap
User-agent:*
Disallow:

> ...

Reviewer: ```suggestion
# ...
```

Author: Why?

Reviewer: Because that’s a comment. All examples mentioning that you may have additional things, is through a comment like // ...

Author: Got it, to keep it valid 👍

The above snippet creates the following routes:

> 

Reviewer: I would change the examples below to an actual `xml` document output.

> The above snippet creates the following routes:

Reviewer: > the above snippet creates the following routes:

Change to 

> the API route above would respond with the following XML:

* https://moxy.studio/some-project
* https://moxy.studio/another-project

---

* https://moxy.studio/old/old-post
* https://moxy.studio/old/ancient-post

---

* https://moxy.studio/new-post/new
* https://moxy.studio/another-new-post/new

> ### Specifying a custom warning function

Reviewer: I think we can remove this. The description in the API section above is sufficient.

```js
setupSitemap({
    baseUrl: 'https://moxy.studio',
    handleWarning: (message) => {
        // do stuff here
    },

> })

Reviewer: ```suggestion
});
```

```

> ### Specifying a custom error function

Reviewer: Same.

```js
setupSitemap({
    baseUrl: 'https://moxy.studio',
    handleError: (error) => {
        // do stuff here
    },
})
```

### Chained dynamic routing

There's also the option of chaining dynamic routes.

This works with pattern matching, and you also receive the previously generated results to assist your mappings.

```js
setupSitemap({
    baseUrl: 'https://app-finder.blabla',
    mapDynamicRoutes: {
        '/[company]': async () => {
            const companies = await fetchCompanies(); // returns ['microsoft', 'apple']

            return companies;
        },
        '/[company]/[apps]': ({ company }) => {
            const apps = fetchCompanyApps(company); // Returns ['vscode', 'vsstudio'] for microsoft and returns ['itunes'] for apple

            return apps;
        },
    },
}),
```

> Assuming that fetchCompanyApps returns `vscode` and `vsstudio` for Microsoft and `itunes` for Apple,

Reviewer: > the above snippet creates the following routes:

Change to 

> the API route above would respond with the following XML:

the above snippet creates the following routes:

> * https://app-finder.blabla/microsoft

Reviewer: Same, print the XML instead.
